text/3844-next-gen-transmute.md

```diff
@@ -0,0 +1,261 @@
+Just add `union_transmute` and change `transmute` as above in `core`.
+
+## Possible implementation approach
+
+Today we already have a `transmute_unchecked` intrinsic in rustc which doesn't
+have a compile-time size check, but *is* still documented as UB if the sizes
+don't match.  That intrinsic can be changed to be defined as the union version,
+perhaps with fallback MIR using that exact definition, and used to implement
+both of the library functions.  Those library functions would probably also
+add some UbChecks (which aren't possible today as `mem::transmute` is a
+re-exported intrinsic, not an actual function).
+
+The change to make `mem::transmute` an ordinary function would need to update
+the existing `check_transmutes` in typeck to instead be a lint that looks for
+calls to `#[rustc_diagnostic_item = "transmute"]` instead.  (That diagnostic
+item already exists.)  For starters the same logic could be used as a
+deny-be-default lint, as the most similar diagnostics, with any splitting done
+separately over time at the discretion of the diagnostics experts.
+
+This should be straight-forward in CTFE and codegen as well.  Once lowered such
+that we have locals for the source and destination, this can be implemented by
+just copying over the number of bytes in the shorter value (plus uninitializing
+the upper bytes in the target, if it's bigger).
+
+For example, in cg_clif the general-case copy currently uses the destination size
+<https://github.com/rust-lang/rust/blob/6d091b2baa33698682453c7bb72809554204e434/compiler/rustc_codegen_cranelift/src/value_and_place.rs#L641>
+but it could use the min of the source and destination size.
+
+In cg_ssa the general case for SSA values actually already supports this
+<https://github.com/rust-lang/rust/blob/6d091b2baa33698682453c7bb72809554204e434/compiler/rustc_codegen_ssa/src/mir/rvalue.rs#L309-L316>
+and thus would just need the earlier "just emit `unreachable` if the sizes don't
+match" check removed to reach it.
+
+There are some other cases that will need more care, like transmuting a larger
+`BackendRepr::Scalar` to a smaller `BackendRepr::Memory` where the current code
+would do a OOB write if unchanged, but nothing particularly troublesome is expected.
+
+The internal changes (to codegen and similar) would probably happen first so they
+could be implemented and tested before doing the publicly-visible switchover.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- The more transmute-related functions we add the more people might feel encouraged
+  to use them, even if we'd rather not.
+- Lots of people don't like post-mono errors, and would rather Rust never have them.
+- This is still massively-unsound, so doesn't solve the biggest problems.
+- Weird crimes using transmute to check sizes without ever actually running the
+  transmute might not get caught by the linting or post-mono check.
+
+
+# Rationale and alternatives
```

Comment:
In the interest of exploring an alternative:

 Could we have `std::mem::transmute` resolve to a different function across editions? Older editions get the typeck-erreors, newer editions get the new `union_transmute` behaviour.

This would technically be breaking in that type equality of the function item of `std::mem::transmute_e2024` and `std::mem::transmute_e2025` would be different. Though I think the only way to actually encounter this in practice would be to have a macro that returns the function item of transmute to a crate on a different edition, which seems unlikely :thinking: 

I think having a coercion from e2025->e2024 transmute would be sound though. Which may even further make that harder to encounter.

Reply:
I don’t think you need any type-level wizardry if you have edition-dependent name resolution, and also a way to override the edition default